src/query/api/v1/handler/prometheus/remote/read.go

```diff
@@ -98,71 +88,80 @@ func newPromReadMetrics(scope tally.Scope) promReadMetrics {
 	}
 }
 
-func (h *PromReadHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+func (h *promReadHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	timer := h.promReadMetrics.fetchTimerSuccess.Start()
+	defer timer.Stop()
+
 	ctx := context.WithValue(r.Context(), handler.HeaderKey, r.Header)
-	logger := logging.WithContext(ctx, h.instrumentOpts)
-	req, rErr := h.parseRequest(r)
+	logger := logging.WithContext(ctx, h.opts.InstrumentOpts())
+	req, fetchOpts, rErr := parseRequest(ctx, r, h.opts)
 	if rErr != nil {
-		xhttp.Error(w, rErr.Inner(), rErr.Code())
-		return
-	}
-
-	timeout, err := prometheus.ParseRequestTimeout(r, h.timeoutOpts.FetchTimeout)
-	if err != nil {
+		err := rErr.Inner()
 		h.promReadMetrics.fetchErrorsClient.Inc(1)
-		xhttp.Error(w, err, http.StatusBadRequest)
-		return
-	}
-
-	fetchOpts, rErr := h.fetchOptionsBuilder.NewFetchOptions(r)
-	if rErr != nil {
-		xhttp.Error(w, rErr.Inner(), rErr.Code())
+		logger.Error("remote read query parse error",
+			zap.Error(err),
+			zap.Any("req", req),
+			zap.Any("fetchOpts", fetchOpts))
+		xhttp.Error(w, err, rErr.Code())
 		return
 	}
 
-	readResult, err := h.read(ctx, w, req, timeout, fetchOpts)
+	cancelWatcher := handler.NewResponseWriterCanceller(w, h.opts.InstrumentOpts())
+	readResult, err := Read(ctx, cancelWatcher, req, fetchOpts, h.opts)
 	if err != nil {
 		h.promReadMetrics.fetchErrorsServer.Inc(1)
 		logger.Error("remote read query error",
 			zap.Error(err),
 			zap.Any("req", req),
-			zap.Duration("timeout", timeout),
 			zap.Any("fetchOpts", fetchOpts))
 		xhttp.Error(w, err, http.StatusInternalServerError)
 		return
 	}
 
+	// NB: if this errors, all relevant headers and information should already
+	// be sent to the writer; so it is not necessary to do anything here other
+	// than increment success/failure metrics.
+	err = WriteSnappyCompressed(w, readResult, logger)
+	if err != nil {
+		h.promReadMetrics.fetchErrorsServer.Inc(1)
+	} else {
+		h.promReadMetrics.fetchSuccess.Inc(1)
+	}
+}
+
+// WriteSnappyCompressed writes snappy compressed results to the given writer.
+func WriteSnappyCompressed(
+	w http.ResponseWriter,
+	readResult ReadResult,
+	logger *zap.Logger,
+) error {
 	resp := &prompb.ReadResponse{
-		Results: readResult.result,
+		Results: readResult.Result,
 	}
 
 	data, err := proto.Marshal(resp)
 	if err != nil {
-		h.promReadMetrics.fetchErrorsServer.Inc(1)
+		// h.promReadMetrics.fetchErrorsServer.Inc(1)
```

Comment:
Can restore this?

Reply:
The metrics will get incremented in the calling function; changed `writeSnappyCompressed` to private since it's not used outside of the caller